Python package building rework

Antelox opened the pull request:
**Your checklist for this pull request**
- [ ] I've documented or updated the documentation of every API function and struct this PR changes.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)

**Detailed description**

- Refactored `setup.py` to remove hacks regarding packaging of wheels for different platforms, improve and cleanup the code
- Updated `README.txt`
- Removed old `Makefile` and `build_wheel.sh` scripts
- Created a new workflow that takes care of building and testing python packages for different platforms/architectures/python versions
- Fixed `cstest_py` positional argument
- Fixed root path detection in `integration_tests.py`
- Added `run_tests.py` script to run all the python unittests
- Added options to specify build type and threads to be used during building
- Refreshed the list of files to be copied into `src` folder for `sdist` archive

**Test plan**

There are 6 tests failing for both Linux and Windows x86 arch.

```
INFO  - Run test: TestInput { arch: mos65xx, options: ['CS_OPT_DETAIL', 'CS_MODE_MOS65XX_W65C02', 'CS_OPT_SYNTAX_MOTOROLA'], addr: 1000, bytes: [ 0x07,0x12,0x27,0x12,0x47,0x12,0x67,0x12,0x87,0x12,0xa7,0x12,0xc7,0x12,0xe7,0x12,0x10,0xfe,0x0f,0x12,0xfd,0x4f,0x12,0xfd,0x8f,0x12,0xfd,0xcf,0x12,0xfd ] }
WARNING - CS_ARCH: Capstone doesn't have the attribute 'mos65xx'
ERROR - Normalized asm-text doesn't match:
decoded:  'bbr0 $12, (null)f0fb60df'
expected: 'bbr0 $12, $1012'

INFO  - TestResult.FAILED

INFO  - Run test: TestInput { arch: mos65xx, options: ['CS_OPT_DETAIL', 'CS_MODE_MOS65XX_65816_LONG_MX', 'CS_OPT_SYNTAX_MOTOROLA'], addr: 1000, bytes: [ 0xa9,0x34,0x12,0xad,0x34,0x12,0xbd,0x34,0x12,0xb9,0x34,0x12,0xaf,0x56,0x34,0x12,0xbf,0x56,0x34,0x12,0xa5,0x12,0xb5,0x12,0xb2,0x12,0xa1,0x12,0xb1,0x12,0xa7,0x12,0xb7,0x12,0xa3,0x12,0xb3,0x12,0xc2,0x00,0xe2,0x00,0x54,0x34,0x12,0x44,0x34,0x12,0x02,0x12 ] }
ERROR - Normalized asm-text doesn't match:
decoded:  'mvn $12, (null)f0fb60df'
expected: 'mvn $12, $34'

INFO  - TestResult.FAILED

INFO  - Run test: TestInput { arch: CS_ARCH_ARM, options: ['CS_MODE_THUMB', 'CS_OPT_DETAIL'], addr: 0, bytes: [ 0x18,0xf8,0x03,0x1e ] }
ERROR - regs_write_count: 2 != 1
INFO  - TestResult.FAILED

INFO  - Run test: TestInput { arch: CS_ARCH_ARM, options: ['CS_MODE_THUMB', 'CS_MODE_MCLASS', 'CS_OPT_DETAIL'], addr: 0, bytes: [ 0x98,0xfc,0x4e,0x08 ] }
ERROR - imm: 4029476313 != 90
INFO  - TestResult.FAILED

INFO  - Run test: TestInput { arch: CS_ARCH_ARM, options: ['CS_MODE_THUMB', 'CS_MODE_V8', 'CS_OPT_DETAIL'], addr: 0, bytes: [ 0x94,0xfd,0x46,0x48 ] }
ERROR - imm: 4029476493 != 270
INFO  - TestResult.FAILED
```


**Closing issues**

I suppose it should close:

- closes https://github.com/capstone-engine/capstone/issues/2445
- closes https://github.com/capstone-engine/capstone/issues/2548
- closes https://github.com/capstone-engine/capstone/issues/1701
- closes https://github.com/capstone-engine/capstone/issues/1292
- closes https://github.com/capstone-engine/capstone/issues/1796

Antelox committed:
- Refactored setup.py to remove hacks regarding packaging of wheels for different platforms, improve and cleanup the code
- Updated README.txt
- Removed old Makefile and build_wheel.sh scripts
- Created a new workflow that takes care of building and testing python packages for different platforms/architectures/python versions

Rot127 commented on .github/workflows/build-wheels-publish.yml:
```diff
@@ -0,0 +1,303 @@
+      matrix:
+        include:
+          # NOTE: Making this to parallelize and speed up workflow
+          # i686 - manylinux
+          - { os: ubuntu-latest, arch: i686, cibw_build: 'cp38-manylinux', cibw_skip: '' }
+          # i686 - musllinux
+          - { os: ubuntu-latest, arch: i686, cibw_build: 'cp38-musllinux', cibw_skip: '' }
+          # x86_64 - manylinux
+          - { os: ubuntu-latest, arch: x86_64, cibw_build: 'cp38-manylinux', cibw_skip: '' }
+          # x86_64 - musllinux
+          - { os: ubuntu-latest, arch: x86_64, cibw_build: 'cp38-musllinux', cibw_skip: '' }
+          # aarch64 - manylinux
+          - { os: ubuntu-latest, arch: aarch64, cibw_build: 'cp38-manylinux', cibw_skip: '' }
+          # aarch64 - musllinux
+          - { os: ubuntu-latest, arch: aarch64, cibw_build: 'cp38-musllinux', cibw_skip: '' }
+          # macos - x86_64
+          - { os: macos-13, arch: x86_64, cibw_build: 'cp38', cibw_skip: '' }
+          # macos - arm64
+          - { os: macos-latest, arch: arm64, cibw_build: 'cp38', cibw_skip: '' }
+          # windows - x86_64
+          - { os: windows-latest, arch: AMD64, cibw_build: 'cp38', cibw_skip: '' }
+          # windows - amd64
+          - { os: windows-latest, arch: x86, cibw_build: 'cp38', cibw_skip: '' }
+          # windows - arm64
+          - { os: windows-latest, arch: ARM64, cibw_build: 'cp39', cibw_skip: '' }
```
We are still on the free plan for Capstone. And would like to stay on it until we manage to get some funding.
This means `MacOS - ARM64` is too expensive (10x the costs of a x86_64 Linux runner last time I checked).
Also you can remove the `Python 3.9` runners. We just ran the oldest and newest Python version and hope in between we are fine as well.

@kabeor How much runtime is used on average currently? If there is still some slack we could include some more.
Or remove the 32bit runners if it is already at the limit.

Antelox replied:
What I wanted to achieve here is leverage as much as possible the matrix parallelism of Github runners, rather having cibuildwheel single plan dealing with multiple versions/platforms. Beside speeding-up aspect, it should also be able to spot issues without stopping at first failed build: for instance if we want to build for manylinux and musllinux, in case manylinux fails we will not see if it will also fail for musllinux since it will exit. I do understand the concern about quota so once we decide how to go ahead I'm available to adapt accordingly.

Regarding the python version basically what I did is to build and test for each push/PR only the lowest python version and all the rest in case of tag release (for a matter of time). You see python 3.9 only for the Windows ARM64 build since cibuildwheel has no py3.8 target for it, that is why. The approach of using lowest/higher I think it makes sense and I can do the change. Even though I would keep the building of the other versions in case of tag release to be sure the wheels uploaded to Pypi are not broken.

Rot127 replied:
@kabeor knows how the quota is currently used. He must give a comment to this.

I am in general a big fan of testing everything on more divers system configurations.
For now it would be nice if you could add some field like `run_on: [push, merge_to_dev, release_create ...]` to each matrix entry. So we can toggle on and off which configuration to run in different cases.

Antelox replied:
the suggested `run_on` things probably can be accomplished in a different PR

Rot127 commented on .github/workflows/build-wheels-publish.yml:
```diff
@@ -0,0 +1,303 @@
+            cstest_py {project}/tests/MC/ &&
+            cstest_py {project}/tests/details/ &&
+            cstest_py {project}/tests/issues/ &&
+            cstest_py {project}/tests/features/
```
Please move them into a script and do some `echo "RUN xxx"` before every command or just do it directly with `set -ex` to print and exit early (no idea what the equivalent on Windows is though).

This way it is easier to spot what failed.

Antelox replied:
not sure worth it having them into separate scripts (per platform). Meanwhile I have pushed some fixes that made simplify the cibuildwheel tests command. The scripts would just add a print and after all you would need to walk the logs anyway, and they already contain a summary of what tests have failed.

Rot127 replied:
Thought about it. But it is worth having them in an extra script, because it can be run locally. Otherwise, if people do changes to these lines, they have to push the file or run `act` or similar to test it. Both is rather time and resource consuming.
So a script is better for testing.

You can move the `python -m pip install {package}/cstest_py` into a separated step. Because they are the same for Windows/Linux.

Antelox replied:
The step of installing `cstest_py` has to be done only after the capstone wheel got built so it will install it and use it for testings. Also in case of Linux build that command will be executed within Docker container

Antelox replied:
Added `run_tests.py` script under `suite` folder and adapted workflow accordingly

Rot127 commented on bindings/python/setup.py:
```diff
@@ -125,85 +115,64 @@ def build_libraries():
+
+    conf = 'Debug' if int(os.getenv('DEBUG', 0)) else 'Release'
+    cmake_args = ['cmake',
+                  '-DBUILD_SHARED_LIBS=ON',
```
Note: After https://github.com/capstone-engine/capstone/pull/2534 it is `CAPSTONE_BUILD_SHARED_LIBS`

Rot127 replied:
You can also disable the build for static libs here.

Antelox replied:
Done

Rot127 commented on bindings/python/setup.py:
Please add your SPDX copyright here. Like this:

```
# SPDX-FileCopyrightText: 2024 Rot127 <unisono@quyllur.org>
# SPDX-License-Identifier: BSD-3
```

Rot127 reviewed (commented):
Looks way way cleaner! Thanks a lot.
I start the CI once for all the jobs. But in general we have to discuss with @kabeor about the possible CI runtimes with this one.

Also, do I see this right? You test only on 3.8 and 3.9 here?
Testing on 3.8 + 3.13 (or whatever is the oldest/newest) is better IMHO.

Rot127 commented:
The failing tests are possibly related to https://github.com/capstone-engine/capstone/issues/2416. Both are clearly 32bit system issues.

Antelox committed:
Added SPDX headers to the setup.py

Antelox committed:
- cstest_py: Fixed positional argument since it doesn't accept a `required` flag. It turns to have a mandatory tests folder path
- integration_tests.py: Use pathlib to determine the required path
- GitHub action: Simplified the tests execution command

Rot127 commented on bindings/python/cstest_py/src/cstest_py/cstest.py:
```diff
@@ -405,34 +404,16 @@ def run_tests(self):
+    parser.add_argument(
+        dest="search_dir",
+        help="Directory to search for .yaml test files.",
+        type=Path,
```
I you remove the default entry, you need to make this one a `required=True`

Antelox replied:
Positional args are required by default. That indeed the actual fix

Antelox replied:
Also this is more clean I think. You install _cstest_py_ as a normal python package then you can run it from whatever you want by specifying the required arg that represents the folder of tests you want to execute. This would remove any implicit assertion, such as stay in git repo and use git cli to derive the tests folder.

Rot127 replied:
Yeah, it is a good change.

Rot127 commented on .github/workflows/build-wheels-publish.yml:
```diff
@@ -0,0 +1,279 @@
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          # NOTE: Making this to parallelize and speed up workflow
+          # i686 - manylinux
+          - { os: ubuntu-latest, arch: i686, cibw_build: 'cp38-manylinux', cibw_skip: '' }
+          # i686 - musllinux
+          - { os: ubuntu-latest, arch: i686, cibw_build: 'cp38-musllinux', cibw_skip: '' }
+          # x86_64 - manylinux
+          - { os: ubuntu-latest, arch: x86_64, cibw_build: 'cp38-manylinux', cibw_skip: '' }
+          # x86_64 - musllinux
+          - { os: ubuntu-latest, arch: x86_64, cibw_build: 'cp38-musllinux', cibw_skip: '' }
+          # aarch64 - manylinux
+          - { os: ubuntu-latest, arch: aarch64, cibw_build: 'cp38-manylinux', cibw_skip: '' }
+          # aarch64 - musllinux
+          - { os: ubuntu-latest, arch: aarch64, cibw_build: 'cp38-musllinux', cibw_skip: '' }
+          # macos - x86_64
+          - { os: macos-13, arch: x86_64, cibw_build: 'cp38', cibw_skip: '' }
+          # macos - arm64
+          - { os: macos-latest, arch: arm64, cibw_build: 'cp38', cibw_skip: '' }
+          # windows - x86_64
+          - { os: windows-latest, arch: AMD64, cibw_build: 'cp38', cibw_skip: '' }
+          # windows - amd64
+          - { os: windows-latest, arch: x86, cibw_build: 'cp38', cibw_skip: '' }
+          # windows - arm64
+          - { os: windows-latest, arch: ARM64, cibw_build: 'cp39', cibw_skip: '' }
```
```suggestion
        include:
          # x86_64 - manylinux
          - { os: ubuntu-latest, arch: x86_64, cibw_build: 'cp38-manylinux', cibw_skip: '' }
          # windows - x86_64
          - { os: windows-latest, arch: AMD64, cibw_build: 'cp38', cibw_skip: '' }
```

Antelox replied:
Done

Antelox committed:
GitHub Actions: Run python 3.8 (lowest) and 3.13 (current highest) for native runners only during testings and the rest during tag release

Antelox committed:
GitHub Action:
- Fixed the cibw_build matrix element
- Added a step to prepare artifact name

Antelox committed:
GitHub Action: Added run_tests.py script to run all tests during CI workflow

Rot127 commented on .github/workflows/build-wheels-publish.yml:
```diff
@@ -0,0 +1,281 @@
+          - { os: ubuntu-latest, arch: aarch64, cibw_build: 'cp311-musllinux*', cibw_skip: '' }
+          - { os: ubuntu-latest, arch: aarch64, cibw_build: 'cp312-musllinux*', cibw_skip: '' }
+          # macos - x86_64
+          - { os: macos-13, arch: x86_64, cibw_build: 'cp*', cibw_skip: '*36* *37* *38* *313*' }
```
Please remove Python 3.6, 3.7. They are EOL and we no longer support them.

Antelox replied:
indeed we skip them :)

Rot127 replied:
Ah sorry, yeah. Was not reading properly.

Rot127 commented on .github/workflows/build-wheels-publish.yml:
```diff
@@ -0,0 +1,281 @@
+name: Build wheels with cibuildwheel
```
```suggestion
name: Build and publish wheels with cibuildwheel
```

Rot127 commented on suite/run_tests.py:
```diff
@@ -0,0 +1,21 @@
+import logging
```
Your SPDX copyright please. Something like this:
```
# Copyright © 2022 Rot127 <unisono@quyllur.org>
# SPDX-License-Identifier: BSD-3
```

Rot127 replied:
Also for the other workflow file please. And you can extend the others as well.

Rot127 commented on .github/workflows/build-wheels-publish.yml:
```diff
@@ -0,0 +1,281 @@
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0  # Optional, use if you use setuptools_scm
+          submodules: true  # Optional, use if you have submodules
```
```suggestion
          fetch-depth: 0
          submodules: true
```

Rot127 commented on bindings/python/MANIFEST.in:
```diff
@@ -1,5 +1,9 @@
+include BUILDING.md
+graft capstone/lib
+graft capstone/include
+global-include *.a
+global-include *.so.*
+global-include *.*lib
+global-include *.dll
```
```suggestion
global-include *.so.*
global-include *.dylib
global-include *.dll
```

Since you rightfully build without static libs.

Rot127 reviewed (changes requested):
Very nice! Thanks a lot for the edits. Just some more small things please.

Also please do a release in your fork on this branch. So we can test if the release workflow works.

Only the actual `publish`/`pypip upload` step should fail.

Thanks a lot for cleaning up all this.

Rot127 commented on .github/workflows/python-publish-release.yml:
```diff
@@ -1,90 +0,0 @@
-      - name: Check binaries (Windows)
-        if: matrix.os == 'windows-latest'
-        run: |
-          python3.exe suite/check_wheel_bin_arch.py ./wheelhouse/
```
Ah sorry. One last thing. Please add this step as well. Those scripts are somewhat dirty. But we had quite some issues that the wheel was build incorrectly, and contained the wrong binary.

Rot127 replied:
Alternatively, we can also check once all the wheels built for a release. And check that the capstone library is built for the correct arch + OS. This is probably better. The check script was a somewhat desperate move from me to have the correct binaries.

Antelox replied:
I guess those scripts where needed before because of the hacks that were present in the `setup.py` + the workflow was not testing them properly. That is not really needed anymore because we do build on specific platforms/architectures and run tests, makes sense?

Rot127 replied:
Ok, nice. Please add the link to the test release later. I'll check them anyways. Last release it was a bad awakening :D

Rot127 commented:
@kabeor This PR opens the question if Capstone supports x86 systems. The tests fail on 32bit systems (at least Windows). I think it is a good idea to support it. Because there are also modern 32bit CPU architectures out there. So having at least a single x86 job is good reduces the chance of bugs also on the other archtectures.

Rot127 commented:
Something is still wrong with the naming:
https://github.com/capstone-engine/capstone/actions/runs/11813575581/job/32923353685?pr=2538

Antelox commented:
> Something is still wrong with the naming: https://github.com/capstone-engine/capstone/actions/runs/11813575581/job/32923353685?pr=2538

That is ugly I know, but it's like that because that job is skipped hence the matrix doesn't get created and finally the placeholders don't get replaced

Rot127 commented on suite/run_tests.py:
```diff
@@ -0,0 +1,21 @@
+for test in tests:
+    logger.info(f'Running {test}')
+    logger.info("#######################")
+    os.system(test)
```
Just recognize this doesn't work. Because the return status of the command is returned here but not handled.
Better use:
```python
subprocess.run(test.split(" "), check=True)
```

Antelox committed:
- Added SPDX headers to the run_tests.py script and to the build-wheels-publish.yml workflow file
- Minor fixes to the workflow as pointed out in the PR review
- Updated MANIFEST.in to reflect the actual libraries built during python wheel creation process
- Use subprocess.run in place of os.system in run_tests.py script

Rot127 commented:
```
   Successfully installed cstest_py-0.1.0 pyyaml-6.0.2
  INFO:tests:Running python D:\a\capstone\capstone/bindings/python/tests/test_all.py
  INFO:tests:#######################
  Traceback (most recent call last):
    File "D:\a\capstone\capstone/bindings/python/tests/test_all.py", line 3, in <module>
      import test_lite
    File "D:\a\capstone\capstone\bindings\python\tests\test_lite.py", line 4, in <module>
      from capstone import *
  ModuleNotFoundError: No module named 'capstone'
  Traceback (most recent call last):
    File "D:\a\capstone\capstone/suite/run_tests.py", line 23, in <module>
      subprocess.run(test.split(" "), check=True)
    File "C:\Users\runneradmin\AppData\Local\pypa\cibuildwheel\Cache\nuget-cpython\python.3.8.10\tools\lib\subprocess.py", line 516, in run
      raise CalledProcessError(retcode, process.args,
  subprocess.CalledProcessError: Command '['python', 'D:\\a\\capstone\\capstone/bindings/python/tests/test_all.py']' returned non-zero exit status 1.
  Error: Command python -m pip install D:\a\capstone\capstone\bindings\python/cstest_py && python D:\a\capstone\capstone/suite/run_tests.py
   failed with code 1. 

```

Rot127 commented:
This one is weird: https://github.com/capstone-engine/capstone/actions/runs/11818841771/job/32927697780?pr=2538
Happens also for this job. Let's see after the review of the others if it is fixed.

Rot127 commented on suite/run_tests.py:
```diff
@@ -0,0 +1,25 @@
+
+import logging
+import subprocess
+import sys
```
```suggestion
```

Rot127 replied:
Ignore it. Oversaw the `sys.executable`

Rot127 reviewed (commented):
Last little nitpick. Then it looks good to me.

I check the build wheels later this day. And will check a last time after kabeor is done.

Rot127 commented on .github/workflows/build-wheels-publish.yml:
```diff
@@ -0,0 +1,284 @@
+          # macos - x86_64
+          - { os: macos-13, arch: x86_64, cibw_build: 'cp*', cibw_skip: '*36* *37* *38* *313*' }
+          # macos - arm64
+          - { os: macos-latest, arch: arm64, cibw_build: 'cp*', cibw_skip: '*36* *37* *38* *313*' }
```
Almost forgot, please add a `arch: universal2` job as well.

Antelox committed:
GitHub Action:
- Run qemu step only if non-native Linux runner
- Added arch:universal2 matrix element for macos-latest runner

Antelox commented:
> This one is weird: https://github.com/capstone-engine/capstone/actions/runs/11818841771/job/32927697780?pr=2538 Happens also for this job. Let's see after the review of the others if it is fixed.

Should be internal GitHub networking issue

Antelox committed:
Python bindings: Refreshed the list of files needed to be copied for sdist archive

Rot127 reviewed (approved):
@aquynh Can you please check how many runner minutes are used currently?
According to [these](https://docs.github.com/en/actions/using-github-hosted-runners/using-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories) Github docs all latest Linux, Windows and MacOS runners (even an M1 instance) are free.

I can't see if this is also the case for all architectures (like Linux + AArch64, Windows + AArch64 or Windows i686). But if all those are free, we should absolutely enable some more runners.

Rot127 commented on .github/workflows/build-wheels-publish.yml:
```diff
@@ -0,0 +1,286 @@
+        include:
+          # NOTE: Making this to parallelize and speed up workflow
+          # i686 - manylinux
+          - { os: ubuntu-latest, arch: i686, cibw_build: 'cp39-manylinux*', cibw_skip: '' }
+          - { os: ubuntu-latest, arch: i686, cibw_build: 'cp310-manylinux*', cibw_skip: '' }
+          - { os: ubuntu-latest, arch: i686, cibw_build: 'cp311-manylinux*', cibw_skip: '' }
+          - { os: ubuntu-latest, arch: i686, cibw_build: 'cp312-manylinux*', cibw_skip: '' }
+          # i686 - musllinux
+          - { os: ubuntu-latest, arch: i686, cibw_build: 'cp39-musllinux*', cibw_skip: '' }
+          - { os: ubuntu-latest, arch: i686, cibw_build: 'cp310-musllinux*', cibw_skip: '' }
+          - { os: ubuntu-latest, arch: i686, cibw_build: 'cp311-musllinux*', cibw_skip: '' }
+          - { os: ubuntu-latest, arch: i686, cibw_build: 'cp312-musllinux*', cibw_skip: '' }
+          # x86_64 - manylinux
```
```suggestion
          # - { os: ubuntu-latest, arch: i686, cibw_build: 'cp39-manylinux*', cibw_skip: '' }
          # - { os: ubuntu-latest, arch: i686, cibw_build: 'cp310-manylinux*', cibw_skip: '' }
          # - { os: ubuntu-latest, arch: i686, cibw_build: 'cp311-manylinux*', cibw_skip: '' }
          # - { os: ubuntu-latest, arch: i686, cibw_build: 'cp312-manylinux*', cibw_skip: '' }
          # i686 - musllinux
          # - { os: ubuntu-latest, arch: i686, cibw_build: 'cp39-musllinux*', cibw_skip: '' }
          # - { os: ubuntu-latest, arch: i686, cibw_build: 'cp310-musllinux*', cibw_skip: '' }
          # - { os: ubuntu-latest, arch: i686, cibw_build: 'cp311-musllinux*', cibw_skip: '' }
          # - { os: ubuntu-latest, arch: i686, cibw_build: 'cp312-musllinux*', cibw_skip: '' }
          # x86_64 - manylinux
```

Rot127 commented on .github/workflows/build-wheels-publish.yml:
```diff
@@ -0,0 +1,286 @@
+          # windows - amd64
+          - { os: windows-latest, arch: AMD64, cibw_build: 'cp*', cibw_skip: '*36* *37* *38* *313*' }
+          # windows - x86
+          - { os: windows-latest, arch: x86, cibw_build: 'cp*', cibw_skip: '*36* *37* *38* *313*' }
```
```suggestion
          # - { os: windows-latest, arch: x86, cibw_build: 'cp*', cibw_skip: '*36* *37* *38* *313*' }
```

Antelox committed:
GitHub Action: Commented out arch:x86 matrix elements

Antelox commented:
> I can't see if this is also the case for all architectures (like Linux + AArch64, Windows + AArch64 or Windows i686). But if all those are free, we should absolutely enable some more runners.

Those are not native runners so it's not the case.
[Here](https://docs.github.com/en/actions/using-github-hosted-runners/using-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories) you can check what are the specs of the runners. Basically all are x64 plus macos-latest one that is AppleSilicon (ARM).

Antelox committed:
GitHub Action: Run qemu step only if non-native Linux runner

Rot127 commented:
@kabeor The run in case of tag creation: https://github.com/Antelox/capstone/actions/runs/11839185045

Antelox committed:
GitHub Action: Minor fixes

Rot127 reviewed (approved):
Please write in the commit message what the fix was for (next time).
Also, provide a link to the Action run of a release with the latest commit. So kabeor can quickly check.

Rot127 commented:
Binaries look good as well:
```
for f in $(ls ./*/capstone/lib/*); do file $f; done
./capstone-6.0.0a1-cp310-cp310-macosx_10_9_universal2/capstone/lib/libcapstone.dylib: Mach-O universal binary with 2 architectures: [x86_64:\012- Mach-O 64-bit x86_64 dynamically linked shared library, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|NO_REEXPORTED_DYLIBS>] [\012- arm64]
./capstone-6.0.0a1-cp310-cp310-manylinux_2_17_aarch64.manylinux2014_aarch64/capstone/lib/libcapstone.so: ELF 64-bit LSB shared object, ARM aarch64, version 1 (SYSV), dynamically linked, BuildID[sha1]=d139511e718f699bc4ff70698f6e5cd7d2c9d26d, not stripped
./capstone-6.0.0a1-cp310-cp310-win_amd64/capstone/lib/capstone.dll: PE32+ executable (DLL) (GUI) x86-64, for MS Windows, 6 sections
./capstone-6.0.0a1-cp310-cp310-win_arm64/capstone/lib/capstone.dll: PE32+ executable (DLL) (GUI) Aarch64, for MS Windows, 6 sections
./capstone-6.0.0a1-cp39-cp39-macosx_10_9_x86_64/capstone/lib/libcapstone.dylib: Mach-O 64-bit x86_64 dynamically linked shared library, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|NO_REEXPORTED_DYLIBS>
./capstone-6.0.0a1-cp39-cp39-macosx_11_0_arm64/capstone/lib/libcapstone.dylib: Mach-O 64-bit arm64 dynamically linked shared library, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|NO_REEXPORTED_DYLIBS>
./capstone-6.0.0a1-cp39-cp39-musllinux_1_2_x86_64/capstone/lib/libcapstone.so: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, BuildID[sha1]=fd9af6ba25a55df236d41a04d6d4776f7b585cab, with debug_info, not stripped
./capstone-6.0.0a1-cp39-cp39-win_amd64/capstone/lib/capstone.dll: PE32+ executable (DLL) (GUI) x86-64, for MS Windows, 6 sections
```

Rot127 commented:
The `sdist` build fails though (on Linux). Please remove the issue from the list above or fix it.
```
Processing ./capstone-6.0.0a1.tar.gz
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Building wheels for collected packages: capstone
  Building wheel for capstone (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Building wheel for capstone (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [69 lines of output]
      /tmp/pip-build-env-jwlqop42/overlay/lib/python3.12/site-packages/setuptools/dist.py:330: InformationOnly: Normalizing '6.0.0alpha1' to '6.0.0a1'
        self.metadata.version = self._normalize_version(self.metadata.version)
      INFO:root:running bdist_wheel
      INFO:root:running build
      INFO:root:running build_py
      INFO:__main__:Building C extensions
      INFO:__main__:Build Directory: /tmp/pip-req-build-dqzt6b7j/src/build_python
      
      -- The C compiler identification is GNU 14.2.1
      -- The CXX compiler identification is GNU 14.2.1
      -- Detecting C compiler ABI info
      -- Detecting C compiler ABI info - done
      -- Check for working C compiler: /usr/bin/cc - skipped
      -- Detecting C compile features
      -- Detecting C compile features - done
      -- Detecting CXX compiler ABI info
      -- Detecting CXX compiler ABI info - done
      -- Check for working CXX compiler: /usr/bin/c++ - skipped
      -- Detecting CXX compile features
      -- Detecting CXX compile features - done
      -- Performing Test SUPPORTS_MU
      -- Performing Test SUPPORTS_MU - Success
      -- Performing Test SUPPORTS_SHADOWING
      -- Performing Test SUPPORTS_SHADOWING - Success
      -- Performing Test SUPPORTS_SUNINIT
      -- Performing Test SUPPORTS_SUNINIT - Failed
      Enabling CAPSTONE_ARM_SUPPORT
      Enabling CAPSTONE_AARCH64_SUPPORT
      Enabling CAPSTONE_M68K_SUPPORT
      Enabling CAPSTONE_MIPS_SUPPORT
      Enabling CAPSTONE_PPC_SUPPORT
      Enabling CAPSTONE_SPARC_SUPPORT
      Enabling CAPSTONE_SYSTEMZ_SUPPORT
      Enabling CAPSTONE_XCORE_SUPPORT
      Enabling CAPSTONE_X86_SUPPORT
      Enabling CAPSTONE_TMS320C64X_SUPPORT
      Enabling CAPSTONE_M680X_SUPPORT
      Enabling CAPSTONE_EVM_SUPPORT
      Enabling CAPSTONE_MOS65XX_SUPPORT
      Enabling CAPSTONE_WASM_SUPPORT
      Enabling CAPSTONE_BPF_SUPPORT
      Enabling CAPSTONE_RISCV_SUPPORT
      Enabling CAPSTONE_SH_SUPPORT
      Enabling CAPSTONE_TRICORE_SUPPORT
      Enabling CAPSTONE_ALPHA_SUPPORT
      Enabling CAPSTONE_HPPA_SUPPORT
      Enabling CAPSTONE_LOONGARCH_SUPPORT
      Enabling CAPSTONE_XTENSA_SUPPORT
      CMake Error: File /tmp/pip-req-build-dqzt6b7j/src/capstone.pc.in does not exist.
      CMake Error at CMakeLists.txt:879 (configure_file):
        configure_file Problem configuring file
      
      
      CMake Error: File /tmp/pip-req-build-dqzt6b7j/src/capstone-config.cmake.in does not exist.
      CMake Error at /usr/share/cmake/Modules/CMakePackageConfigHelpers.cmake:501 (configure_file):
        configure_file Problem configuring file
      Call Stack (most recent call first):
        CMakeLists.txt:884 (configure_package_config_file)
      
      
      CMake Error: File /tmp/pip-req-build-dqzt6b7j/src/cmake_uninstall.cmake.in does not exist.
      CMake Error at CMakeLists.txt:924 (configure_file):
        configure_file Problem configuring file
      
      
      -- Configuring incomplete, errors occurred!
      gmake: Makefile: No such file or directory
      gmake: *** No rule to make target 'Makefile'.  Stop.
      error: [Errno 2] No such file or directory: 'libcapstone.so.6'
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for capstone
Failed to build capstone
ERROR: Could not build wheels for capstone, which is required to install pyproject.toml-based projects


```

Antelox committed:
Python bindings: Added missing .in pattern when collecting src files for sdist archive

Antelox commented:
> The `sdist` build fails though (on Linux). Please remove the issue from the list above or fix it.
> 
> ```
> Processing ./capstone-6.0.0a1.tar.gz
>   Installing build dependencies ... done
>   Getting requirements to build wheel ... done
>   Preparing metadata (pyproject.toml) ... done
> Building wheels for collected packages: capstone
>   Building wheel for capstone (pyproject.toml) ... error
>   error: subprocess-exited-with-error
>   
>   × Building wheel for capstone (pyproject.toml) did not run successfully.
>   │ exit code: 1
>   ╰─> [69 lines of output]
>       /tmp/pip-build-env-jwlqop42/overlay/lib/python3.12/site-packages/setuptools/dist.py:330: InformationOnly: Normalizing '6.0.0alpha1' to '6.0.0a1'
>         self.metadata.version = self._normalize_version(self.metadata.version)
>       INFO:root:running bdist_wheel
>       INFO:root:running build
>       INFO:root:running build_py
>       INFO:__main__:Building C extensions
>       INFO:__main__:Build Directory: /tmp/pip-req-build-dqzt6b7j/src/build_python
>       
>       -- The C compiler identification is GNU 14.2.1
>       -- The CXX compiler identification is GNU 14.2.1
>       -- Detecting C compiler ABI info
>       -- Detecting C compiler ABI info - done
>       -- Check for working C compiler: /usr/bin/cc - skipped
>       -- Detecting C compile features
>       -- Detecting C compile features - done
>       -- Detecting CXX compiler ABI info
>       -- Detecting CXX compiler ABI info - done
>       -- Check for working CXX compiler: /usr/bin/c++ - skipped
>       -- Detecting CXX compile features
>       -- Detecting CXX compile features - done
>       -- Performing Test SUPPORTS_MU
>       -- Performing Test SUPPORTS_MU - Success
>       -- Performing Test SUPPORTS_SHADOWING
>       -- Performing Test SUPPORTS_SHADOWING - Success
>       -- Performing Test SUPPORTS_SUNINIT
>       -- Performing Test SUPPORTS_SUNINIT - Failed
>       Enabling CAPSTONE_ARM_SUPPORT
>       Enabling CAPSTONE_AARCH64_SUPPORT
>       Enabling CAPSTONE_M68K_SUPPORT
>       Enabling CAPSTONE_MIPS_SUPPORT
>       Enabling CAPSTONE_PPC_SUPPORT
>       Enabling CAPSTONE_SPARC_SUPPORT
>       Enabling CAPSTONE_SYSTEMZ_SUPPORT
>       Enabling CAPSTONE_XCORE_SUPPORT
>       Enabling CAPSTONE_X86_SUPPORT
>       Enabling CAPSTONE_TMS320C64X_SUPPORT
>       Enabling CAPSTONE_M680X_SUPPORT
>       Enabling CAPSTONE_EVM_SUPPORT
>       Enabling CAPSTONE_MOS65XX_SUPPORT
>       Enabling CAPSTONE_WASM_SUPPORT
>       Enabling CAPSTONE_BPF_SUPPORT
>       Enabling CAPSTONE_RISCV_SUPPORT
>       Enabling CAPSTONE_SH_SUPPORT
>       Enabling CAPSTONE_TRICORE_SUPPORT
>       Enabling CAPSTONE_ALPHA_SUPPORT
>       Enabling CAPSTONE_HPPA_SUPPORT
>       Enabling CAPSTONE_LOONGARCH_SUPPORT
>       Enabling CAPSTONE_XTENSA_SUPPORT
>       CMake Error: File /tmp/pip-req-build-dqzt6b7j/src/capstone.pc.in does not exist.
>       CMake Error at CMakeLists.txt:879 (configure_file):
>         configure_file Problem configuring file
>       
>       
>       CMake Error: File /tmp/pip-req-build-dqzt6b7j/src/capstone-config.cmake.in does not exist.
>       CMake Error at /usr/share/cmake/Modules/CMakePackageConfigHelpers.cmake:501 (configure_file):
>         configure_file Problem configuring file
>       Call Stack (most recent call first):
>         CMakeLists.txt:884 (configure_package_config_file)
>       
>       
>       CMake Error: File /tmp/pip-req-build-dqzt6b7j/src/cmake_uninstall.cmake.in does not exist.
>       CMake Error at CMakeLists.txt:924 (configure_file):
>         configure_file Problem configuring file
>       
>       
>       -- Configuring incomplete, errors occurred!
>       gmake: Makefile: No such file or directory
>       gmake: *** No rule to make target 'Makefile'.  Stop.
>       error: [Errno 2] No such file or directory: 'libcapstone.so.6'
>       [end of output]
>   
>   note: This error originates from a subprocess, and is likely not a problem with pip.
>   ERROR: Failed building wheel for capstone
> Failed to build capstone
> ERROR: Could not build wheels for capstone, which is required to install pyproject.toml-based projects
> ```

Should be fixed with commit: https://github.com/capstone-engine/capstone/pull/2538/commits/82fe006a60eb4035932a0a52c0733a8b8f7238ba

kabeor reviewed (approved):
LGTM, Thank you!
